cpg_infra/billing_aggregator/driver.py

```diff
@@ -151,6 +152,45 @@ def create_source_archive(
             opts=pulumi.ResourceOptions(replace_on_changes=['*']),
         )
 
+    def setup_gcp_cost_control(self):
+        """
+        Create the gcp cost control cloud function to cut off billing when
+        it exceeds the budget
+        """
+        region = self.config.gcp.region
+        service_account = self.config.billing.gcp_cost_controls.machine_account
+        slack_channel = self.config.billing.gcp_cost_controls.slack_channel
+        pubsub_topic_name = self.config.billing.gcp_cost_controls.pubsub_topic
+
+        # Create source archive
+        source_archive = self.create_source_archive(
+            'billing-gcp-cost-report-source-code',
+            self.source_bucket.name,
+            os.path.join(os.path.dirname(__file__), 'gcp_cost_control'),
+        )
+
+        # Deploy Cloud Function
+        function = gcp.cloudfunctionsv2.Function(
```

Comment:
Does this work? I would have thought this would need the `build_config` and `service_config` stuff that is in `create_cloud_function`? Or is this meant to be calling `self.create_cloud_function`?

Reply:
Great pickup. You're right! I'd originally used the create_cloud_function until I realised it would be too hard to modify to create a pubsub triggered function. Forgot to wrap the build config parameters correctly.

Reply:
Interesting that our type checking / linting didn't pick anything up, I guess the gcp cloudfunctions method doesn't have typing info attached.